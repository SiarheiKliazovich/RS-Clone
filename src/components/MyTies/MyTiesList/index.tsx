import { FunctionComponent } from "react";
import { useDispatch, useSelector } from "react-redux";

import {
  myTieDeleted,
  myTiesFetching,
  myTiesFetched,
  myTiesFetchingError,
} from "../../../actions";
import { deleteTie, getTiesByUserId } from "../../../services/apiTies";
import { IAuthReducer } from "../../../interfaces/authReducer";
import { IMyTiesListComponent } from "../../../interfaces/myTiesListComponent";
import { useTranslation } from "react-i18next";

const MyTiesList: FunctionComponent<IMyTiesListComponent> = ({ ties }) => {
  const authStore = useSelector((state: IAuthReducer) => state.auth);
  const { user } = authStore;

  const dispatch = useDispatch();

  const { t } = useTranslation("dataLang");

  const deleteItem = async (id: string): Promise<void> => {
    try {
      await deleteTie(id);
      dispatch(myTieDeleted(id));
      dispatch(myTiesFetching());
      const ties = await getTiesByUserId(user._id);
      dispatch(myTiesFetched(ties));
    } catch {
      dispatch(myTiesFetchingError());
    }
  };

  return (
    <>
      <div className="my-ties-list__wrapper">
        {ties.map((item, i) => {
          return (
            <div key={i} className="my-ties-list__item">
              <div className="item__image">
                <img src={item.image} alt="" />
              </div>
              <div className="item__text">
                <div className="item__name">
                  {t("myTies.name")}: {item.name}
                </div>
                <div className="item__price">
                  {t("myTies.price")}: {item.price}$
                </div>
                <button
                  className="item__button"
                  onClick={() => deleteItem(item._id)}
                >
                  {t("myTies.delete")}
                </button>
              </div>
            </div>
          );
        })}
      </div>
    </>
  );
};

export default MyTiesList;
