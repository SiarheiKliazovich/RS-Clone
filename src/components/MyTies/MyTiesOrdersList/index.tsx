import { FunctionComponent } from "react";

import { IOrdersListComponent } from "../../../interfaces/ordersListComponent";
import { useTranslation } from "react-i18next";

const MyTiesOrdersList: FunctionComponent<IOrdersListComponent> = ({
  orders,
}) => {
  const { t } = useTranslation("dataLang");

  return (
    <div className="market-block__products">
      <div className="row__products">
        {orders.map((order, i) => {
          return (
            <div key={i} className="tie__product">
              <div className="tie__products_img">
                <img src={order.image} alt="" className="products_img" />
              </div>
              <div className="tie__products_discription">
                <div className="tie__products_name">
                  {t("myTiesOrders.date")}: {order.date.slice(0, 10)}
                </div>
                <div className="tie__products_price">
                  {t("myTiesOrders.sum")}: {order.price}$
                </div>
              </div>
            </div>
          );
        })}
      </div>
    </div>
  );
};

export default MyTiesOrdersList;
