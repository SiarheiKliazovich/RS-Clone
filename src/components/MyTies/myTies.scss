@import "../../sass/variables";
@import "../../sass/mixins";

.my-ties__wrapper > {
    box-sizing: border-box;
}

.my-ties__wrapper {
    border-radius: 5px;
    padding-top: 40px;
    padding-bottom: 40px;
    box-shadow: $box-shadow;
}

.my-ties__header {
    display: flex;
    justify-content: center;
}

.my-ties-list__wrapper {
    display: flex;
    justify-content: center;
    flex-wrap: wrap;
    margin: 10px auto 0;
    gap: 20px;
    min-height: calc(100vh - 669px);
}

.my-ties-list__item {
    display: flex;
    justify-content: space-between;
    border: 1px solid $font-color-dark;
    overflow: hidden;
    border-radius: 5px;
    margin-bottom: 20px;
    height: 145px;    

    .item__image {
        img {
            width: 100px;
        }
    }
    
    .item__text {
        width: 160px;
        display: flex;
        flex-direction: column;
        justify-content: center;
        align-items: center;
        margin-right: 10px;
    }
    
    .item__name,
    .item__price {
        margin-bottom: 10px;
    }
    
    .item__button {
        width: 80px;
        cursor: pointer;
        border-radius: 5px;
        border: none;
        color: $font-color-white;
        padding: 5px;
    }
}

.my-ties-orders-list__wrapper {
    display: flex;
    justify-content: center;
    flex-wrap: wrap;
    gap: 20px;
    margin-top: 10px;

    @include media-small-mobile {
        gap: 10px;
    }
}

.my-ties-orders-list__item {
    display: flex;
    justify-content: space-between;
    border: 1px solid $font-color-dark;
    overflow: hidden;
    border-radius: 5px;
    margin-bottom: 10px;
    height: 115px; 
    width: 300px;

    .item__image {
        img {
            width: 80px;
        }
    }

    .item__text {
        display: flex;
        flex-direction: column;
        justify-content: center;
        margin: 0 10px;
    }

    .item__date {
        margin-bottom: 10px;
    }
}
