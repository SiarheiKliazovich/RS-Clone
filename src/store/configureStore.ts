/* eslint-disable @typescript-eslint/ban-ts-comment */
import { applyMiddleware, createStore } from "redux";
import thunkMiddleware from "redux-thunk";

import loggerMiddleware from "./logger";
import rootReducer from "../reducers";

const configureStore = () => {
  const middlewares = [loggerMiddleware, thunkMiddleware];
  const middlewareEnhancer = applyMiddleware(...middlewares);

  const store = createStore(rootReducer, middlewareEnhancer);

  // @ts-ignore:next-line
  if (process.env.NODE_ENV !== "production" && module.hot) {
    // @ts-ignore:next-line
    module.hot.accept("./../reducers", () => store.replaceReducer(rootReducer));
  }

  return store;
};

export default configureStore;
