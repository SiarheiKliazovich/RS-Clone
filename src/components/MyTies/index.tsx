import { FunctionComponent, useState, useRef, useEffect } from "react";
import { useSelector, useDispatch } from "react-redux";
import { Link } from "react-router-dom";
import classNames from "classnames";
import { useTranslation } from "react-i18next";

import {
  myTiesFetching,
  myTiesFetched,
  myTiesFetchingError,
  myOrdersFetching,
  myOrdersFetched,
  myOrdersFetchingError,
} from "../../actions";
import { lightTheme, nightTheme } from "../../data/constants";
import { IStore } from "../../interfaces/store";
import Profile from "../Profile";
import Header from "../Header";
import Footer from "../Footer";
import Hover from "../Hover";
import useOnClickOutside from "../../hook/useOnClickOutside";
import navMenu from "../../data/navmenu";
import { ILangReducer } from "../../interfaces/langReducer";
import { IAuthReducer } from "../../interfaces/authReducer";
import { getTiesByUserId } from "../../services/apiTies";
import { getOrdersBySellerId } from "../../services/apiOrders";
import Spinner from "../Spinner";
import { IMyTiesReducer } from "../../interfaces/myTiesReducer";
import { IMyOrdersReducer } from "../../interfaces/myOrdersReducer";
import ErrorMessage from "../ErrorMessage";
import MyTiesList from "./MyTiesList";

import "./myTies.scss";

const MyTies: FunctionComponent = () => {
  const [activeButton, setActiveButton] = useState<string>("1");

  const interfaceSettings = useSelector((state: IStore) => state.appInterface);
  const [open, setOpen] = useState(false);
  const {
    accentColor,
    isSidebarFixed,
    isSidebarAccentMode,
    isNavbarNightMode,
    isProfileShow,
    isNightMode,
  } = interfaceSettings;
  const backgroundColor = isNightMode
    ? nightTheme.background.element
    : lightTheme.background.element;

  const { lang } = useSelector((state: ILangReducer) => state.langReducer);
  const authStore = useSelector((state: IAuthReducer) => state.auth);
  const { user } = authStore;
  const myTiesStore = useSelector(
    (state: IMyTiesReducer) => state.myTiesReducer
  );
  const { tieLoadingStatus, ties } = myTiesStore;

  const myOrdersStore = useSelector(
    (state: IMyOrdersReducer) => state.myOrdersReducer
  );
  const { ordersLoadingStatus, orders } = myOrdersStore;

  const ref = useRef<HTMLDivElement>(null);
  useOnClickOutside(ref, () => setOpen(!open), open);

  const list = navMenu.find((c) => c.lang === lang)!;

  const currentURL = window.location.pathname;

  const { t } = useTranslation("dataLang");

  const dispatch = useDispatch();

  const getDataList = async (n: string) => {
    console.log(n);
    try {
      if (n === "1") {
        dispatch(myTiesFetching());
        const ties = await getTiesByUserId(user._id);
        dispatch(myTiesFetched(ties));
      } else {
        dispatch(myOrdersFetching());
        const orders = await getOrdersBySellerId(user._id);
        dispatch(myOrdersFetched(orders));
      }
    } catch {
      if (n === "1") {
        dispatch(myTiesFetchingError());
      } else {
        dispatch(myOrdersFetchingError());
      }
    }
  };

  useEffect(() => {
    getDataList("1");
  }, []);

  const toggleHandler = (e: React.MouseEvent<HTMLButtonElement>) => {
    const id = e.currentTarget.dataset.id;
    if (typeof id === "string") {
      setActiveButton(id);
    }
  };

  const spinner = tieLoadingStatus === "loading" ? <Spinner /> : null;

  return (
    <>
      <Header
        accentColor={accentColor}
        isNavbarNightMode={isNavbarNightMode}
        isButtonVisible={false}
      >
        <div className="nav__hamburger_wrapper" ref={ref}>
          <span className="nav__hamburger" onClick={() => setOpen(!open)} />
        </div>
        <div
          className={classNames("nav-sidebar", { sidebaropen: open })}
          style={{
            backgroundColor: isNavbarNightMode
              ? backgroundColor
              : accentColor.static,
          }}
        >
          <div className="nav-sidebar-wrapper">
            <ul className="list-nav-sidebar">
              {list.data.map((item) => (
                <li className="nav-sidebar-item" key={item.name}>
                  {currentURL === item.path ? (
                    <span className="active-link">{item.name}</span>
                  ) : (
                    <Link className="nav-sidebar-link" to={item.path}>
                      {item.name}
                    </Link>
                  )}
                </li>
              ))}
            </ul>
          </div>
        </div>
      </Header>
      <Profile
        accentColor={accentColor}
        isProfileShow={isProfileShow}
        isSidebarFixed={isSidebarFixed}
        isSidebarAccentMode={isSidebarAccentMode}
        isNavbarNightMode={isNavbarNightMode}
        isNightMode={isNightMode}
      />
      <div className="my-ties__wrapper" style={{ backgroundColor }}>
        <div className="my-ties__header">
          <Hover>
            <button
              style={{ backgroundColor: accentColor.static }}
              className={`users__header-button ${
                activeButton == "1" ? "users__header-button_active" : ""
              }`}
              data-id="1"
              onClick={(e) => {
                toggleHandler(e);
                getDataList(e.currentTarget.dataset.id!);
              }}
            >
              {t("myTies.myTies")}
            </button>
          </Hover>
          <Hover>
            <button
              style={{ backgroundColor: accentColor.static }}
              className={`users__header-button ${
                activeButton == "2" ? "users__header-button_active" : ""
              }`}
              data-id="2"
              onClick={(e) => {
                toggleHandler(e);
                getDataList(e.currentTarget.dataset.id!);
              }}
            >
              {t("myTies.mySales")}
            </button>
          </Hover>
        </div>
        {spinner}
        {tieLoadingStatus === "idle" && activeButton === "1" ? (
          <MyTiesList ties={ties} activeButton={activeButton} />
        ) : null}
        {tieLoadingStatus === "error" && activeButton === "1" ? (
          <ErrorMessage />
        ) : null}
        {/* {(ordersLoadingStatus === "idle"  && activeButton === "2") ? (
        <OrdersList ties={ties} activeButton={activeButton} />
      ) : null} */}
        {ordersLoadingStatus === "error" && activeButton === "2" ? (
          <ErrorMessage />
        ) : null}
      </div>

      <Footer accentColor={accentColor} isNavbarNightMode={isNavbarNightMode} />
    </>
  );
};

export default MyTies;
