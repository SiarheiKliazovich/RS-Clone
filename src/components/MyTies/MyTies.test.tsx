import { render } from "@testing-library/react";
import { BrowserRouter as Router } from "react-router-dom";
import configureStore from "redux-mock-store";
import { Provider } from "react-redux";
import MyTies from "./";

const mockStore = configureStore([]);

describe("MyTies component", () => {
  // eslint-disable-next-line
  let store: any;

  beforeEach(() => {
    store = mockStore({
      appInterface: {
        accentColor: {
          static: "#000",
          hover: "#fff",
        },
      },
      auth: {
        user: {
          role: "USER",
        },
      },
      myTiesReducer: {
        tieLoadingStatus: "load",
      },
      myOrdersReducer: {
        ordersLoadingStatus: "load",
      },
    });
  });

  test("renders correctly", () => {
    const { container } = render(
      <Provider store={store}>
        <Router>
          <MyTies />
        </Router>
      </Provider>
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
