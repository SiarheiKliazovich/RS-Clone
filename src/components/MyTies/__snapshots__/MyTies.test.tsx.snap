// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MyTies component renders correctly 1`] = `
<div
  class="tiemarket-wrapper"
>
  <div
    class="my-ties__wrapper"
    style="background-color: rgb(246, 246, 246);"
  >
    <div
      class="container"
    >
      <div
        class="market-block"
        style="background-color: rgb(255, 255, 255);"
      >
        <div
          class="my-ties__header"
        >
          <button
            class="users__header-button users__header-button_active"
            data-id="1"
            style="background-color: rgb(255, 255, 255);"
          >
            myTies.myTies
          </button>
          <button
            class="users__header-button"
            data-id="2"
            style="background-color: rgb(0, 0, 0);"
          >
            myTies.mySales
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
